subsys/net/lib/zperf/zperf_shell.c

```diff
@@ -178,6 +178,137 @@ static int parse_ipv4_addr(const struct shell *sh, char *host, char *port,
+	if (argc >= 3) {
+		if (IS_ENABLED(CONFIG_NET_IPV6) && !IS_ENABLED(CONFIG_NET_IPV4)) {
+			ret = net_addr_pton(AF_INET6, argv[2], &sin6_addr);
+			if (ret < 0) {
+				goto ipv6_err;
+			}
+
+			if (net_ipv6_is_addr_unspecified(&sin6_addr)) {
+				goto ipv6_err;
+			}
+
+			memcpy(&param->sin6_addr, &sin6_addr,
+			       sizeof(struct in6_addr));
+
+			return 0;
+		}
+
+		if (IS_ENABLED(CONFIG_NET_IPV4) && !IS_ENABLED(CONFIG_NET_IPV6)) {
+			ret = net_addr_pton(AF_INET, argv[2], &sin_addr);
+			if (ret < 0) {
+				goto ipv4_err;
+			}
+
+			if (net_ipv4_is_addr_unspecified(&sin_addr)) {
+				goto ipv4_err;
+			}
+
+			memcpy(&param->sin_addr, &sin_addr,
+			       sizeof(struct in_addr));
+
+			return 0;
+		}
+
+		if (IS_ENABLED(CONFIG_NET_IPV6) && IS_ENABLED(CONFIG_NET_IPV4)) {
+			if (net_addr_pton(AF_INET6, argv[2], &sin6_addr) < 0) {
+				if (net_addr_pton(AF_INET, argv[2],
+						  &sin_addr) < 0) {
+					goto ipv4_err;
+				}
+
+				if (net_ipv4_is_addr_unspecified(&sin_addr)) {
+					goto ipv4_err;
+				}
+
+				memcpy(&param->sin_addr, &sin_addr,
+				       sizeof(struct in_addr));
+			} else {
+				if (net_ipv6_is_addr_unspecified(&sin6_addr)) {
+					goto ipv6_err;
+				}
+
+				memcpy(&param->sin6_addr, &sin6_addr,
+				       sizeof(struct in6_addr));
+			}
+			return 0;
+		}
+ipv4_err:
+		shell_fprintf(sh, SHELL_WARNING, "Invalid IPv4 address %s\n",
+			      argv[2]);
+		return -ENOEXEC;
+
+ipv6_err:
+		shell_fprintf(sh, SHELL_WARNING, "Invalid IPv6 address %s\n",
+			      argv[2]);
+		return -ENOEXEC;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		if (MY_IP6ADDR && strlen(MY_IP6ADDR)) {
+			/* Use Setting IP */
+			ret = zperf_get_ipv6_addr(MY_IP6ADDR, MY_PREFIX_LEN_STR,
+						  &sin6_addr);
+			if (ret < 0) {
+				shell_fprintf(sh, SHELL_WARNING,
+					      "Unable to set IPv6.\n");
+				goto use_existing_ipv6;
+			}
+			memcpy(&param->sin6_addr, &sin6_addr,
+			       sizeof(struct in6_addr));
+		} else {
+			/* Use existing IP */
+			const struct in6_addr *addr;
+use_existing_ipv6:
+			addr = zperf_get_default_if_in6_addr();
+			if (!addr) {
+				shell_fprintf(sh, SHELL_WARNING,
+					      "Unable to get IPv6 by default.\n");
+				return -ENOEXEC;
+			}
+			memcpy(&param->sin6_addr, addr,
+			       sizeof(struct in6_addr));
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		if (MY_IP4ADDR && strlen(MY_IP4ADDR)) {
+			/* Use Setting IP */
+			ret = zperf_get_ipv4_addr(MY_IP4ADDR, &sin_addr);
+			if (ret < 0) {
+				shell_fprintf(sh, SHELL_WARNING,
+					      "Unable to set IPv4.\n");
+				goto use_existing_ipv4;
+			}
+			memcpy(&param->sin_addr, &sin_addr,
+			       sizeof(struct in_addr));
+		} else {
+			/* Use existing IP */
+			const struct in_addr *addr;
+use_existing_ipv4:
+			addr = zperf_get_default_if_in4_addr();
+			if (!addr) {
+				shell_fprintf(sh, SHELL_WARNING,
+					      "Unable to get IPv4 by default.\n");
+				return -EINVAL;
+			}
+			memcpy(&param->sin_addr, addr, sizeof(struct in_addr));
+		}
+	}
```

Comment:
This moves part of the `zperf` library functionality into the zperf shell component. Note, that zperf shell is optional, the library has a standalone API that can be used by some apps.

IMO this code should remain where it was, just extend it with a check, to configure the address only if the one provided with `zperf_download_params` is unspecified. zperf shell should only focus on parsing the arguments, and filling out the `zperf_download_params` structure.